test/PushPullTests.cpp

```diff
@@ -0,0 +1,460 @@
+      std::vector<std::shared_ptr<Pull>> toReturn;
+      for (int i = 0; i < howManyPulls ; i ++) {
+         if (howManyPushs == 1 && howManyPulls == 1) {
+            // one to one, use default behavior (bind)
+            LOG(DEBUG) << "one to one pull setup";
+            toReturn.push_back(std::make_shared<Pull>(location));
+         } else if (howManyPushs > 1) {
+            LOG(DEBUG) << "many to one pull setup";
+            // many pushes to one pull, use default behavior (bind)
+            toReturn.push_back(std::make_shared<Pull>(location));
+         } else if (howManyPulls > 1) {
+            LOG(DEBUG) << "one to many pull setup";
+            // many pulls to one push, connect instead of bind
+            toReturn.push_back(std::make_shared<Pull>(location, true));
+         } else if (howManyPushs == 0 && howManyPulls == 1) {
+            LOG(DEBUG) << "one pull setup";
+            // one pull, use default behavior (bind)
+            toReturn.push_back(std::make_shared<Pull>(location));
+         }
+      }
+      return toReturn;
+   }
+   std::vector<std::thread> ListenWithPulls(std::vector<std::shared_ptr<Pull>>& pulls,
+                                            std::function<void(std::shared_ptr<Pull>)> pullLambda) {
+      std::vector<std::thread> threads;
+      for (auto& pull:pulls) {
+         LOG(DEBUG) << "Creating Pull thread";
+         threads.push_back(std::thread(pullLambda, pull));
+      }
+      return threads;
+   }
+   std::vector<std::thread> ShootWithPushs(std::vector<std::shared_ptr<Push>>& pushs,
+                        std::function<void(std::shared_ptr<Push>)> pushLambda) {
+      std::vector<std::thread> threads;
+      for (auto& push:pushs) {
+         LOG(DEBUG) << "Creating push thread";
+         threads.push_back(std::thread(pushLambda, push));
+      }
+      return threads;
+   }
+   void FireStrings(int howManyPushs,
+                    int howManyShots,
+                    int howManyPulls,
+                    const std::string& location,
+                    const std::string& data) {
+      LOG(DEBUG) << "FireStrings";
+      auto pushList = CreatePushs(howManyPushs, howManyPulls, location);
+      auto pullList = CreatePulls(howManyPushs, howManyPulls, location);
+      auto pushLambda = [&](std::shared_ptr<Push> threadGun) {
+         for (int i = 0; i < howManyShots / howManyPushs; i++) {
+            auto thread_id = std::this_thread::get_id();
+            LOG(DEBUG) << thread_id << " about to fire ";
+            threadGun->Send(data);
+            LOG(DEBUG) << thread_id << " fired shot: ";
+         }
+      };
+      auto pullLambda = [&](std::shared_ptr<Pull> threadPull) {
+         auto thread_id = std::this_thread::get_id();
+         for (int i = 0; i < howManyShots / howManyPulls; i++) {
+            LOG(DEBUG) << thread_id << " waiting for push shot";
+            auto queueString = threadPull->GetString();
```

Comment:
Please look and compare ```FireStrings``` with ```FireVectors``` with ```FirePointers```. There are about  2 lines that differ between each function and each is at least 30 lines. 

Please see if you can merge these together to one function instead of 3 almost identical